### ISSUE TITLE
fpprintprec do not round bfloat correctly(another case)

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#84]
Imported from SourceForge on 2024-07-01 18:05:28
Created by **[pellicula](https://sourceforge.net/u/pellicula/)** on 2017-04-17 17:15:59
Original: https://sourceforge.net/p/maxima/bugs/3301

---

(%i1) display2d:false;
(%o1) false
(%i2) fpprintprec:4;
(%o2) 4
(%i3) 9.9999e0;
(%o3) 10.0 <- correct
(%i4) 9.9999b0;
(%o4) 1.0b0 <- not correct

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:29
Created by **[pellicula](https://sourceforge.net/u/pellicula/)** on 2017-04-17 17:26:05
Original: https://sourceforge.net/p/maxima/bugs/3301/#0e54

---

and...

(%i1) display2d:false;
(%o1) false
(%i2) fpprintprec:4;
(%o2) 4
(%i3) string(9.9999e0);
(%o3) "10.0" <- correct
(%i4) string(9.9999b0);
(%o4) "1.0b0" <- not correct
(%i5) printf(false, "~h", 9.9999b0);
(%o5) "1.0" <- not correct

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:32
Created by **[pellicula](https://sourceforge.net/u/pellicula/)** on 2017-04-17 17:29:25
Original: https://sourceforge.net/p/maxima/bugs/3301/#db30

---

cf. https://sourceforge.net/p/maxima/bugs/3213/

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:34
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-01-28 20:26:03
Original: https://sourceforge.net/p/maxima/bugs/3301/#e3c8

---

Another example:

fpprintprec:6$
0.9999999b0 =>  1.0b-1   <<<!!! should be 1.0b0.

In fact, you get this incorrect answer for bfloat(1-10^-i) for i in 4..16 and fpprintprec=i-2 or i-1.

~~~
for i:4 thru 16 do
   for fpprintprec:2 thru i-1 do 
      print(i,fpprintprec,bfloat(1-10^-i))
~~~
Actually, anything from ``1-0.5*10^-i`` to ``1-5*10^-i`` hits this bug.

Tested on Maxima 5.45.1 SBCL 2.0.0 Windows

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:36
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-01-28 20:46:51
Original: https://sourceforge.net/p/maxima/bugs/3301/#51ac

---

- **labels**:  --> fpprintprec
- **Priority**: 5 --> 9

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:38
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-01-30 16:35:27
Original: https://sourceforge.net/p/maxima/bugs/3301/#7833

---

Here is some proposed code to fix ``fpformat``.

The existing code was poorly formatted (incorrect indendation) and very messy (reusing the variable name ``l`` for several different things). So I reformatted it and changed some variable names to understand it better before applying my patch. It can certainly be cleaned up further, but my main goal was to fix this bug.

Please read and try out the code and let me know what you think.





Attachments:
- https://sourceforge.net/p/maxima/bugs/_discuss/thread/ea7e1215/7833/attachment/fpformat.lisp

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:40
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-04 23:04:29
Original: https://sourceforge.net/p/maxima/bugs/3301/#2196

---

It looks like this bug was actually fixed by Ray Toy a while ago (commit d1976ae dated 2021-09-24). I don't see the buggy behavior with a current build.

Version 5.45.1 is dated 2021-06-20 so it doesn't include the bug fix. 

I think it's still meaningful to clean up the code and include new test cases, including any specific ones mentioned here.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:42
Created by **[rtoy](https://sourceforge.net/u/rtoy/)** on 2023-02-04 23:33:57
Original: https://sourceforge.net/p/maxima/bugs/3301/#2df4

---

Huh.  I don't remember doing that at all.  And here's a link to the commit [d1976a], just to make it easier to see.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:44
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-05 18:46:36
Original: https://sourceforge.net/p/maxima/bugs/3301/#3ca4

---

- **labels**: fpprintprec --> fpprintprec, bigfloat
- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-01 18:05:47
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-05 18:46:36
Original: https://sourceforge.net/p/maxima/bugs/3301/#b079

---

Closing this ticket as fixed by commit d1976ae. We can continue to work on FPFORMAT but anyway this bug is fixed.
